elastic/kibana
# [HTTP/OAS] HTTP API `/api/oas`

## Summary

Adds a new core-owned experimental HTTP API: `GET /api/oas` with the following query params for filtering:

- `pluginId` - request a specific plugin's OAS. **Note**: this is the opaque id, e.g. `@kbn/maps-plugin`
- `pathStartsWith` - request OAS for paths that start with the given string

The current rationale is that this will provide sufficient scoping to allow developers and docs maintainers to request only the OAS they need (since Kibana's surface area is HUGE).

Also added a new experimental config: `server.oas.enabled: <boolean>`. Set this to `true` to play around with this feature locally.

## How to test

1. Add `server.oas.enabled: true` to `kibana.dev.yml`
2. Start Kibana with `yarn start --no-base-path`
3. `curl -uelastic:changeme http://localhost:5601/api/oas\?pathStartsWith\=/api/status | jq`
4. You should have a JSON representation of an OAS doc containing the `/api/status` endpoint!

<details>

<summary>example</summary>

```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "Kibana HTTP APIs",
    "version": "0.0.0"
  },
  "servers": [
    {
      "url": "http://localhost:5601"
    }
  ],
  "paths": {
    "/api/status": {
      "get": {
        "responses": {},
        "parameters": [
          {
            "in": "header",
            "name": "elastic-api-version",
            "schema": {
              "type": "string",
              "enum": [
                "2023-10-31"
              ],
              "default": "2023-10-31"
            }
          },
          {
            "name": "v7format",
            "in": "query",
            "required": false,
            "schema": {
              "type": "boolean"
            }
          },
          {
            "name": "v8format",
            "in": "query",
            "required": false,
            "schema": {
              "type": "boolean"
            }
          }
        ],
        "operationId": "/api/status#0"
      }
    }
  },
  "components": {
    "schemas": {},
    "securitySchemes": {
      "basicAuth": {
        "type": "http",
        "scheme": "basic"
      },
      "apiKeyAuth": {
        "type": "apiKey",
        "in": "header",
        "name": "Authorization"
      }
    }
  },
  "security": [
    {
      "basicAuth": []
    }
  ]
}

```

</details>

## Risks

* The work to generate OAS can be expensive (for both CPU and memory). To mitigate, the endpoint will only be registered provided a specific config and we are using concurrency control (`Semaphore`) to allow only one of these operations to be handled at a time.

## Future work

* We must start the long tail work of furnishing endpoints with missing response schemas. I intend to do so for the `/api/status` route next
